includes/data-events/connectors/ga4/class-ga4.php

```diff
@@ -0,0 +1,185 @@
+	 * @param int   $client_id ID of the client that triggered the event.
+	 *
+	 * @throws \Exception If the event is invalid.
+	 * @return void
+	 */
+	public static function handle_reader_logged_in( $timestamp, $data, $client_id ) {
+
+		$params = array_merge(
+			self::get_default_params(),
+			[ 'user_id' => $data['user_id'] ]
+		);
+
+		$event = new Event( 'reader_login', $params );
+		self::send_event( $event, $client_id );
+	}
+
+	/**
+	 * Get the default parameters that are added to all events
+	 *
+	 * @return array
+	 */
+	private static function get_default_params() {
+		$params = [
+			'logged_in' => is_user_logged_in() ? 'yes' : 'no',
+		];
+
+		if ( class_exists( 'Automattic\Jetpack\Device_Detection' ) ) {
+			self::log( 'Detecting device using user agent: ' . $_SERVER['HTTP_USER_AGENT'] ?? '(empty)' ); // phpcs:ignore
+			$device                = Device_Detection::is_phone() ? 'phone' : ( Device_Detection::is_tablet() ? 'tablet' : 'desktop' );
+			$params['device_type'] = $device;
+		}
+		return $params;
+	}
+
+	/**
+	 * Gets the credentials for the GA4 API.
+	 *
+	 * @return array
+	 */
+	private static function get_credentials() {
+		$api_secret     = get_option( 'ga4_api_secret' );
+		$measurement_id = get_option( 'ga4_measurement_id' );
+		return compact( 'api_secret', 'measurement_id' );
+	}
+
+	/**
+	 * Gets the API URL for GA4
+	 *
+	 * @return WP_Error|string
+	 */
+	public static function get_api_url() {
+		$credentials    = self::get_credentials();
+		$api_secret     = $credentials['api_secret'];
+		$measurement_id = $credentials['measurement_id'];
+		if ( ! $api_secret || ! $measurement_id ) {
+			return new WP_Error( 'missing_credentials', 'Missing GA4 API credentials.' );
+		}
+		return add_query_arg(
+			[
+				'api_secret'     => $api_secret,
+				'measurement_id' => $measurement_id,
+			],
+			'https://www.google-analytics.com/mp/collect'
+		);
+	}
+
+	/**
+	 * Sends an event to GA4.
+	 *
+	 * @param Event  $event The event object.
+	 * @param string $user_id User identifier. Use Reader_Activation::get_client_id().
+	 * @throws \Exception If the credentials are missing.
+	 * @return void
+	 */
+	private static function send_event( Event $event, $user_id = null ) {
+
+		$url = self::get_api_url();
+
+		if ( is_wp_error( $url ) ) {
+			throw new \Exception( $url->get_error_message() );
+		}
+
+		$client_id = self::extract_cid_from_cookies();
+
+		$payload = [
+			'client_id' => $client_id,
+			'events'    => [
+				$event->to_array(),
+			],
+		];
+
+		if ( $user_id ) {
+			$payload['user_id'] = $user_id;
+		}
+
+		wp_remote_post(
+			$url,
+			[
+				'body' => wp_json_encode( $payload ),
+			]
+		);
+
+		self::log( sprintf( 'Event sent - %s', $event->get_name() ) );
+
+	}
+
+	/**
+	 * Extracts the Client ID from the _ga cookie
+	 *
+	 * If the cookie is not found, it will be created
+	 *
+	 * @return string
+	 */
+	private static function extract_cid_from_cookies() {
```

Comment:
Right before the `Event sent - reader_login` log I'm getting this warning:

```
PHP Warning:  Cannot modify header information - headers already sent in newspack-plugin/includes/data-events/connectors/ga4/class-ga4.php on line 166
```

Reply:
I'm guessing it's not happy with the `setcookie()` part

Reply:
I've removed it. If the cookie is not set, we use a anonymous user, just as we are already doing [here](https://github.com/Automattic/newspack-plugin/blob/add%2Fga4-connector/includes/oauth/class-google-services-connection.php#L109)